### ISSUE TITLE
P12.4: Unify initial generation and repair with masked program denoising

### ISSUE THREAD

**Issue by kiwi1005** [kiwi1005/CadC26#28]
Parent: #24

Depends on: #26 KEEP; #27 when preplaced support is required

Branch: [`research/model-first-e2e-floorplanner`](https://github.com/kiwi1005/CadC26/tree/research/model-first-e2e-floorplanner)

# Hypothesis

The same shared model can generate a floorplan from scratch and refine a partial/decoded solution by treating both as masked reconstruction of one FloorplanProgram. Iterative masked program denoising should improve decoded QoR over one-shot generation without handing global search back to an algorithm.

# Core idea

Use one `GenerationState` with three task modes:

```text
FULL_MASK        -> direct initial solution
PARTIAL_PROGRAM  -> partial reconstruction
DECODED_RESIDUAL -> regenerate low-confidence program tokens after exact compile
```

The model is responsible for deciding regenerated topology/shape tokens. The algorithm only applies masks, compiles each model output once, verifies and selects among K model samples.

# Structured mask curriculum

Do not add Gaussian `(x,y)` noise. Corrupt program structure:

```text
M0: mask 10–20% rank/aspect tokens
M1: mask one local block group
M2: mask 50% movable tokens
M3: mask 75–90% movable tokens
M4: mask 100% movable tokens (full generation)
```

Optional later corruptions, only when directly measured:

- swap local permutation ranks;
- detach a contact/subtree relation;
- corrupt aspect class;
- mask low-confidence boundary-order tokens.

Every derivative of one clean source stays in one split.

# Model changes

- partial positive/negative assignment embeddings;
- known/masked bits;
- noise-level embedding;
- task token;
- per-token confidence;
- optional round embedding.

No new global solver, router or expert framework.

# Inference

Compare:

```text
one-shot
4 rounds
8 rounds
12 rounds
```

At each round:

1. predict all masked program tokens;
2. keep a scheduled fraction of highest-confidence assignments/aspects;
3. remask only low-confidence tokens;
4. compile only at the final round for the first experiment;
5. later compile intermediate rounds only if it gives a decisive signal.

Generate K=4 programs through sample embeddings; do not beam-search topology outside the model.

# Training

Start from the #26/#27 checkpoint.

Loss is computed only on masked tokens plus existing clean relation/shape/outline losses. Mix mask levels in a fixed curriculum:

```text
first: M0–M2
then:  M0–M4
```

Reuse P11 Contact corruption only as a low-mask auxiliary sample type. It must not dominate the training distribution.

# Experiment

Use the same frozen source-held-out bucket and compiler as #26/#27. Compare one-shot versus iterative generation under equal K and report:

- compile success;
- hard feasibility;
- official-style QoR;
- area/HPWL and soft debt;
- token-confidence calibration;
- rounds and runtime;
- whether improvement comes from model regeneration or fallback;
- full-mask versus partial-mask generalization.

# Decision

## KEEP

- iterative generation improves source-held-out decoded QoR by at least 5% over the frozen one-shot model, or produces clear unique wins;
- compile success/hard feasibility do not regress materially;
- runtime remains bounded and model-forward dominated;
- partial-program repair and full-mask generation both work through the same representation;
- no algorithmic topology search is introduced.

## MODIFY

Only for one identified confidence/mask schedule problem.

## REJECT

- multiple rounds only reduce training loss but not decoded QoR;
- one-shot is already equal or better;
- iterative path becomes an algorithm-driven search loop;
- low-mask repair works but full-mask generation collapses.

# Tests

- deterministic mask generation;
- source split isolation;
- confidence schedule;
- FULL_MASK/PARTIAL/RESIDUAL state distinction;
- one-state masked overfit;
- full-mask compile path;
- one-shot versus iterative gate report.

# Next experiment

KEEP -> #29 add only measured constraint/residual heads. REJECT -> keep one-shot model and continue only if #26/#27 QoR is already valuable.
